### PR TITLE
Only set an experiment cookie and send an enrollment event on the client

### LINKED ISSUE #10188: Update experiment code to deal with anonymous page caching
Because experiments rely on cookies, we need to make sure that our anonymous page caching scheme doesn't interfere with the running of experiments, and vice versa.

Some things to consider:
1. We need the correct cookie to be sent for each user in the experiment.
2. Ideally we should be able to cache a page for an experiment user (as opposed to disabling caching for anyone enrolled in an experiment).
3. We cannot `Vary` on a particular cookie value.

Some suggestions from our chat in Slack:
1. We could have a single cookie that lists all experiments a user in enrolled in, along with the corresponding variant. Then nginx can use that value as part of the cache key.
2. We should always clean up cookies after an experiment is over. Currently we just set them to expire in 30 days, but we should add code to delete them (or rather, update the single cookie mentioned above) when the experiment is no longer active.
3. If we change an experiment (i.e., change some of the UX associated with a variant) we need to ensure that we assign a new, unique id for the experiment. This is something we would generally do anyway, but it becomes critical now, because of caching.
4. We could use a scheme similar to what addons-server does for activity logs by adding all experiment ids to a file, keeping old ones around, having a list with them all and adding a test to make sure the list only contains unique items.

@diox Can you think of anything we discussed that I haven't captured?

### PR DESCRIPTION
**Pull request by bobsilverberg**
Fixes #10188 and re-enables the experiment on dev

### CODE REVIEW

**Comment by bobsilverberg** [comment]
@willdurand Interestingly, this required very little changes to code, and the components will all behave in the optimal manner because they are designed to render the "current" version when `variant` is `null`, and in this case the current version is the less disruptive version.

File: src/amo/withExperiment.js [hunk 1]
```diff
@@ -26,6 +26,12 @@ import { getDisplayName } from 'amo/utils';
  *        in the experiment, assign a percentage to the special
  *        `NOT_IN_EXPERIMENT` variant.
  *
+ *  Note: The component will initially be rendered with `variant === null`.
+ *        Therefore the version of the component that will be the least
+ *        disruptive should be generated when `variant === null`. Once a
```
Comment (willdurand) [inline src/amo/withExperiment.js:31]: We should probably explain that experiments are only being executed on the client before mentioning "disruptive versions".

File: src/amo/withExperiment.js [hunk 2]
```diff
@@ -167,49 +173,52 @@ export const withExperiment = ({
...
-      const experimentsToStore = { ...registeredExperiments };
-
-      // Clear any disabled experiments from the cookie.
-      let cleanupNeeded = false;
-      for (const experimentId of Object.keys(registeredExperiments)) {
-        if (!_isExperimentEnabled({ _config, id: experimentId })) {
-          delete experimentsToStore[experimentId];
-          cleanupNeeded = true;
+      // We only want to run this code if we are on the client.
+      if (!_config.get('server')) {
```
Comment (willdurand) [inline src/amo/withExperiment.js:177]: Can we use `componentDidMount()` instead of the `constructor()` or is it too late?

Comment (bobsilverberg) [reply src/amo/withExperiment.js:177]: We can, and it does seem to work, but I'm not sure how to write a test for the case where we're not running on the client. I guess we don't do that for any of our components that use `componentDidMount`, and just rely on the fact that it doesn't run on the server. 

This all seems a bit fragile to me, and non-explicit, but I guess it's the way we've always done it, and we rely on that behaviour in a lot of places. I wonder what we're going to do if we ever move to hooks? I guess that's an issue for the future AMO team.

I can change it and remove the tests that verify it doesn't run on the server.

Comment (willdurand) [reply src/amo/withExperiment.js]: do you think we could write an integration test like we do here? https://github.com/mozilla/addons-frontend/blob/364dafc88412c3527cd49c9345404ccb64c68e20/tests/unit/amo/server/test_base.js#L557

Comment (bobsilverberg) [reply src/amo/withExperiment.js]: Maybe. I can give it a try. I guess one might ask why we're doing it for this and not for any other instances where we rely on `componentDidMount` not getting called on the server. I know I'm the one suggesting we should have a test for this, but I'm also anticipating that question from you. 😛

**Review by eviljeff — APPROVED** [review]
seems reasonable but I'd defer to @willdurand

**Comment by bobsilverberg** [comment]
@willdurand I guess this is waiting on your r?. I don't mind trying to write an integration test as you suggest, but perhaps we can make that a follow-up as I can see it being quite time consuming and I'd like to get this back up on dev.

**Review by willdurand — APPROVED** [review]

**Pull request merged**